### PR TITLE
Improve performance in __CF_IsBridgedObject/__CF_IsCFObject

### LINKED ISSUE #1490: CFRetain and CFRelease could be expensive for bridged classes
_CFLocaleCopyCurrentGuts for example spends most of its time in CFRetain because CFLocale is bridged to NSLocale.  @DHowett-MSFT has partial fix for this [here](https://github.com/DHowett-MSFT/WinObjC/commits/201609-nscftype).

### PR DESCRIPTION
**Pull request by ms-jihua**
- __CF_IsCFObject() calls __CF_IsBridgedObject()
 - Previously, __CF_IsBridgedObject() would do an O(n^2) comparison against the entire runtime class table.
   (O(n) traversal of table * O(n) check of inheritance tree)
 - These two functions are used in CF memory management functions, such as CFRetain/Release,
   and due to their prevalence, took up substantial portions of CPU time.
 - Changed bridged classes to adopt an empty 'bridged object' protocol when they are bridged/added to the class table.
   This protocol can be quickly checked against, and obviates the need to traverse the table.
 - Removed some redundant _cfisa checks in _CF_IsCFObject

Fixes #1490

### CODE REVIEW

**Comment by ms-jihua** [comment]
Benchmarks: 
image

File: Frameworks/CoreFoundation/Base.subproj/CFInternal.h [hunk 1]
```diff
@@ -696,28 +706,23 @@ CF_INLINE bool __CF_IsBridgedObject(CFTypeRef obj) {
...
-    }
-
-    return false;
+    // Check if the object has the 'bridged object' protocol
+    // NOTE: because our class_conformsToProtocol() checks based only on name, 
+    // it is possible to spoof a bridged object by stating an object conforms to a protocol with the same name.
+    // However, this would not grant access to anything that could not be accessed by more conventional means
+    // (ie: overriding an existing CF/Foundation class)
+    // and as such would not constitute an escalation of privilege.
+    return [object conformsToProtocol:__CFRuntimeGetBridgeProtocol()];
```
Comment (DHowett-MSFT) [inline Frameworks/CoreFoundation/Base.subproj/CFInternal.h]: Prefer `class_conformsToProtocol(object->_cfisa, ...)` -- that will remove Objective-C from purely C++ CFInternal.h consumers.

Comment (DHowett-MSFT) [inline Frameworks/CoreFoundation/Base.subproj/CFInternal.h]: (this header is shared externally -- CoreGraphics, CoreText, etc. use it)

File: Frameworks/CoreFoundation/Base.subproj/CFInternal.h [hunk 2]
```diff
@@ -696,28 +706,23 @@ CF_INLINE bool __CF_IsBridgedObject(CFTypeRef obj) {
...
-    for (unsigned int i = 0; i < __CFRuntimeClassTableSize; i++) {
-        if ((__CFRuntimeObjCClassTable[i] != 0) && 
-            ((object->_cfisa == __CFRuntimeObjCClassTable[i]) ||
-            ([object isKindOfClass:(Class)(__CFRuntimeObjCClassTable[i])]))) {
-            return true;
-        }
-    }
-
-    return false;
+    // Check if the object has the 'bridged object' protocol
```
Comment (aballway) [inline Frameworks/CoreFoundation/Base.subproj/CFInternal.h]: nit: should have WINOBJC: here

Comment (ms-jihua) [inline Frameworks/CoreFoundation/Base.subproj/CFInternal.h]: This is already a WINOBJC: function, so not necessary

**Comment by ms-jihua** [comment]
Benchmarks after handling NSCFType properly:

image

**Comment by ms-jihua** [comment]
Before/after: 
image

**Comment by ms-jihua** [comment]
Note that for these changes to propagate, it may be necessary to do a rebuild - our incremental build seems to be bad at picking up changes to CF_INLINE functions, though nothing seems to break.

**Pull request merged**